File: 075/sort-colors.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var sortColors = function (nums) {
    let len = nums.length;
    if (!len) return;

    let temp = nums[0];
    nums[0] = 1;

    let [i, j, k] = [0, 1, len-1];

    while (j <= k) {
        if (nums[j] < 1) {
            let swap = nums[i];
            nums[i] = nums[j];
            nums[j] = swap;
            i++;
            j++;
        } else if (1 < nums[j]) {
            let swap = nums[j];
            nums[j] = nums[k];
            nums[k] = swap;
            k--;
        } else {
            j++;
        }
    }

    if (temp == 0) {
        nums[i] = temp;
    } else if (temp == 2) {
        nums[k] = temp;
    }
    return;
};

File: 075/sort-colors.go
package main

import "fmt"

func sortColors(nums []int) {
	length := len(nums)
	if length == 0 {
		return
	}

	temp := nums[0]
	nums[0] = 1

	i, j, k := 0, 1, length-1
	for j <= k {
		switch {
		case nums[j] < 1:
			nums[i], nums[j] = nums[j], nums[i]
			i++
			j++
		case 1 < nums[j]:
			nums[j], nums[k] = nums[k], nums[j]
			k--
		default:
			j++
		}
	}

	switch temp {
	case 0:
		nums[i] = temp
	case 2:
		nums[k] = temp
	}

	return
}

func main() {
	nums := []int{2, 0, 2, 1, 1, 0}
	fmt.Println(nums)
	sortColors(nums)
	fmt.Println(nums)
}
